### ISSUE TITLE
Not able to connect to RabbitMQ with a virtual host configured.

### ISSUE THREAD

**Issue by ashfaqDell** [knative-extensions/eventing-rabbitmq#1216]
**Describe the bug**
I am not able to connect to an external rabbitMQ cluster which is accessible through my C# code. In the external connection string configuration, I have not found a place to provide rabbitMQ virtual host. I think it is mandatory for me to provide that to connect to the external rabbitMQ instance as I have access only to that virtual host. 

 **Expected behavior**
I should see a connection established to rabbitMQ

 **To Reproduce**
1. Create a virtual host in a rabbitMQ cluster
2. Create a user and give that user access to that virtual user
3. Provide connection details in rabbitMQ secret file 
4. Use RabbitMQSource to connect to rabbitMQ by providing the previously created secret.

 **Knative release version**
latest

**Comment by gabo1208** [comment]
Hey @ashfaqDell !! Thank you for catching that detail, you are absolutely right we are missing docs regarding that specially in the Broker side. Here I can tell you where to put it for the RabbitMQ Broker and Source:
For the Source:
```
apiVersion: sources.knative.dev/v1alpha1
kind: RabbitmqSource
metadata:
  name: rabbitmq-source-vhost
  namespace: {{ .namespace }}
spec:
  rabbitmqClusterReference:
    name: rabbitmqc
    namespace: {{ .namespace }}
  rabbitmqResourcesConfig:
    exchangeName: "logs"
    queueName: "test"
    vhost: "test-vhost"
  sink:
    ref:
      apiVersion: v1
      kind: Service
      name: recorder
```

For the Broker it goes in the BrokerConfig resource, and the Trigger uses the same vhost as the Broker:
```
apiVersion: eventing.knative.dev/v1alpha1
kind: RabbitmqBrokerConfig
metadata:
  name: rabbitmq-broker-config
  namespace: {{ .namespace }}
spec:
  vhost: "test-vhost"
  rabbitmqClusterReference:
    name: rabbitmqc
    namespace: {{ .namespace }}
```

And if you want to contribute the docs, you are welcome, if you can't then next week I'll be updating them on this regard. Thanks for the concise Report and use case :)!

**Comment by ashfaqDell** [comment]
**Thank you very much. I will be glad to contribute to the documentation. I have one more question. In the secrets file I am providing something like below.**

apiVersion: v1
kind: Secret
metadata:
  name: rabbitmq-secret-credentials
  namespace: test
stringData:  
  username: mydomain\myuser
  password: mypassword
  uri: https://test.com:15672

**And my source is like below**

`apiVersion: sources.knative.dev/v1alpha1
kind: RabbitmqSource
metadata:
  name: rabbitmq-source
  namespace: test
spec:
  rabbitmqClusterReference:
    connectionSecret:
      name: rabbitmq-secret-credentials
       # a secret with your external cluster credentials
  rabbitmqResourcesConfig:    
    exchangeName: "eventing-rabbitmq-source"
    queueName: "eventing-rabbitmq-source"
    vhost: "myHost"
    
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1
      kind: Broker
      name: test
      namespace: default`

**Our rabbitMQ server uses default certs and is on SSL. In my C# code I need to provide URL like below.**

`ConnectionFactory factory = new ConnectionFactory();
factory.UserName = "mydomain\myuser";
factory.Password = "mypassword";
factory.VirtualHost = "myHost";
factory.Uri = new Uri("AMQPS://test.com:8071");
IConnection conn = factory.CreateConnection();`

**This works perfectly fine. But through Knative even after providing the vhost, I am unable to connect. I feel this might be because I need to tell Knative to use AMQPS. Can you suggest me here please.**

**Comment by gabo1208** [comment]
Of that we do have an example, check the: https://github.com/knative-extensions/eventing-rabbitmq/blob/main/samples/external-cluster/200-secret.yaml

tl;dr. Set the `port` field in the connection secret:
```
apiVersion: v1
kind: Secret
metadata:
  name: rabbitmq-secret-credentials
  namespace: external-cluster-sample
# This is just a sample, don't use it this way in production
stringData:
  username: $EXTERNAL_RABBITMQ_USERNAME
  password: $EXTERNAL_RABBITMQ_PASSWORD
  uri: $EXTERNAL_RABBITMQ_URI:$HTTP_PORT # # https://example.com:12345, example.com:12345, rabbitmqc.namespace:15672
  port: $AMQP_PORT # 5672 default
```

The `amqps` part is infered from the `https` url

**Comment by ashfaqDell** [comment]
**I tried putting the port, but unfortunately, it's not connecting. As I mentioned before, my C# program has no issue connecting to the same cluster. Below is my updated secret.**

apiVersion: v1
kind: Secret
metadata:
name: rabbitmq-secret-credentials
namespace: test
stringData:
username: mydomain\myuser
password: mypassword
uri: https://test.com:15672/
port: "8071"

**Comment by gabo1208** [comment]
Any message that you are receiving? Also try to use the port as an int and not a string just to test

**Comment by ashfaqDell** [comment]
I did some more tests. Looks like the vhost is definitely the problem. I changed the port to the default port in the cluster. And ran the same test. Even with the default port, the vhost is not connecting. In order to further dig deeper, I removed the vhost from the rabbitmqSource file and added admin credentials of the rabbitmq cluster. I was able to establish a connection using knative. So vhost is definitely a problem. 

With a C# program with or without vhost I was able to connect to the same cluster. But with knative I am able to connect only without vhost.  The secret file and rabbitMQSource file which is working for me are below.

apiVersion: v1
kind: Secret
metadata:
name: rabbitmq-secret-credentials
namespace: test
stringData:
username: admin
password: admin123
uri: https://test.com:15672/

**RabbitMQ Source**

`apiVersion: sources.knative.dev/v1alpha1
kind: RabbitmqSource

metadata:
name: rabbitmq-source
namespace: test

spec:
rabbitmqClusterReference:

connectionSecret:
name: rabbitmq-secret-credentials

rabbitmqResourcesConfig:
exchangeName: "eventing-rabbitmq-source"
queueName: "eventing-rabbitmq-source"

sink:
ref:
apiVersion: eventing.knative.dev/v1
kind: Broker
name: test
namespace: default`

**Comment by gabo1208** [comment]
Awesome, let me dig a bit into this :)! I'll have some time by end of the week

**Comment by gabo1208** [comment]
Do you have the rabbitmq cluster logs??

**Comment by ashfaqDell** [comment]
Hi there. Sorry for the late reply. I found the logs.
' parse "amqps://americas\\userr:password1@paas-rmq-svc-dev01.us.dell.com:8071": invalid port ":password1" after host`. 

Looks like it is trying to think of the password as the port. It might be because of the domain name **americas** in the username

**Comment by ashfaqDell** [comment]
I was also wondering how to configure Dead letter queue in RabbitMQ source? it is not clear from documentation. Can you please reply once you get a chance.

**Comment by gabo1208** [comment]
> I was also wondering how to configure Dead letter queue in RabbitMQ source? it is not clear from documentation. Can you please reply once you get a chance.

Right now, there is no DLQ support for RabbitMQ Sources. Is on the issue list but not a high priority one

**Comment by gabo1208** [comment]
Ok that log was helpful, probably tomorrow I'll have a fix ready. Next tuesday will be on the next patch release.

Let me know if you need any specific version patch with the fix

**Comment by ashfaqDell** [comment]
Our rabbitMQ cluster is integrated with the corporate directory. We can not add users directly. When we add users we have to pull it from the active directory. All users exist with a domain in our directory.

**Comment by ashfaqDell** [comment]
Quick Question? What is the best way to handle failures in case of RabbitMQSource. After multiple retries what happens to the event if the sink is still down? Are you at least sending a NACK to rabbitMQ cluster? So that RabbitMQ itself can send that message to a dead letter queue?

**Comment by gabo1208** [comment]
Yes we send nacks to rabbitmq and the failure response to the event producer =)

So you could do the dead lettering yourself via rabbitmq policies or any other way you want to handle those messages.
In my opinion depends on your needs, but it never hurts to have a DLQ, if it is not necessary then handle the error on the producer.

**Comment by gabo1208** [comment]
Found the culprit, the field caSecretName is missing from the `rabbitmq-secret-credentials` `Secret`. Try to fix it and let me know how it goes @ashfaqDell 

Also fixed the rabbitm mtls and external cluster samples because they were kind of broker, so now everything should be fine now :). The best news is that there is no new release, nor anything to change on the code for the fix hehe

**Comment by ashfaqDell** [comment]
what is the value of caSecret? Is this rabbitmq-ca as shown below?

**kubectl -n rabbitmq-system create secret generic rabbitmq-ca --from-file=ca.crt=$CA_PATH**

I followed your documentation and I have my new files as below

**Secret**
apiVersion: v1
kind: Secret
metadata:
name: rabbitmq-secret-credentials
namespace: test
stringData:
username: mydomain\myuser
password: mypassword
uri: https://test.com:15672/
port: "8071"
caSecretName: rabbitmq-ca

and 

**RabbitMQ Source**

`apiVersion: sources.knative.dev/v1alpha1
kind: RabbitmqSource

metadata:
name: rabbitmq-source
namespace: test

spec:
rabbitmqClusterReference:

connectionSecret:
name: rabbitmq-secret-credentials

rabbitmqResourcesConfig:
exchangeName: "eventing-rabbitmq-source"
queueName: "eventing-rabbitmq-source"

sink:
ref:
apiVersion: eventing.knative.dev/v1
kind: Broker
name: test
namespace: default`

**No luck.. Same error**

 kubectl get events -n external
LAST SEEN   TYPE      REASON          OBJECT                                                                MESSAGE
12s         Warning   InternalError   rabbitmqsource/rabbitmq-source                                        parse "amqps:// mydomain\\myuser:mypassword@test.com:8071": invalid port ":mypassword" after host
36s         Normal    Killing         pod/rabbitmqsource-rabbitmq-source-d9c8a885d3a0a6295b65d9226bf68vpg   Stopping container receive-adapter

**Comment by gabo1208** [comment]
Yes the `ca-secret` is the same in the whole sample

There is a weird space between `amqps://` and `mydomain\` 🤔 try two things just to check:
- Remove the space
- Scape the `\` using double `\\`

Just to see if it works. I'll try to test the go client with an user with the `\` character this weekend to see if it's the client fault

**Comment by gabo1208** [comment]
And with just one slash is failing too?

**Comment by ashfaqDell** [comment]
Yes, the messages below are with single and double slashes respectively.

`5m46s       Warning   InternalError   rabbitmqsource/rabbitmq-source   parse "amqps://mydomain\\myuser:mypassword@test.com:8071": invalid port ":mypassword" after host
3s          Warning   InternalError   rabbitmqsource/rabbitmq-source   parse "amqps://mydomain\\\\myuser:mypassword@test.com:8071": invalid port ":mypassword" after host  `

**Comment by ashfaqDell** [comment]
Another question : I used local rabbitmq instance and posted the message to rabbitmq-source which has failer service configured. Based on the samples you have given for the rabbitmq retry configuration, I can see the failer logs responding correctly. I observed that after all the retries, there is no NACK for the rabbitMQ server. I don't see message appearing in the configured DLQ after all retries are done. When I use the same configuration in my C# program to knack a message, I can see messages re-appearing in DLQ. So my conclusion is that either failer program is not properly handing off the message back to the knative, or the knative itself is not knacking the message.

**Comment by gabo1208** [comment]
The Source is nacking, but it sounds weird, maybe create another issue, so this one does not get overloaded and let's discuss there?
Could be a good start to share how did you set up the DLQ for the Source :)

**Comment by gabo1208** [comment]
Going back to the bug regarding user with "\" @ashfaqDell, you cannot use "/" instead?

I think I have a possible fix but just to make sure it's necesary.

Basically we are using url.Parse to create the rabbitmqUrl from the secret/cluster reference, but just parsing it could be enough
